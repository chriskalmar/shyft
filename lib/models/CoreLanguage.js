
import {
  dataTypes,
  Entity,
} from '../../';


const {
  DataTypeString,
} = dataTypes



export const CoreLanguage = new Entity({
  name: 'language',
  domain: 'core',
  description: 'A language',

  // indexing: {
  //   unique: [
  //     [ 'name' ],
  //     [ 'iso_code' ],

  attributes: {

    name: {
      type: DataTypeString,
      description: 'The name of the language',
      maxLength: 30,
      required: true
    },

    nativeName: {
      type: DataTypeString,
      description: 'The native name of the language',
      maxLength: 30,
      required: true
    },

    isoCode: {
      type: DataTypeString,
      description: 'ISO code of the language',
      pattern: '^[a-z]+$',
      minLength: 2,
      maxLength: 2,
      required: true
    },
  }

})

