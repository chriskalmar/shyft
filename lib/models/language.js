
'use script';

module.exports = {

  domain: 'core',

  entities: [

    {
      name: 'language',
      description: 'a language',

      cascadeOnDelete: true,

      attributes: [

        {
          name: 'name',
          type: 'string',
          description: 'Name',
          maxLength: 30,
          required: true
        },

        {
          name: 'native_name',
          type: 'string',
          description: 'Name',
          maxLength: 30,
          required: true
        },

        {
          name: 'iso_code',
          type: 'string',
          description: 'ISO 639-1 code of the language',
          maxLength: 2,
          required: true
        },

      ],

      indexing: {
        unique: [
          [ 'name' ],
          [ 'iso_code' ],
        ]
      }

    }

  ]

}
