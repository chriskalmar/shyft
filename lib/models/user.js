
'use script';

module.exports = {

  domain: 'core',

  entities: [

    {
      name: 'user',
      description: 'a user',

      isUserTable: true,
      cascadeOnDelete: true,

      attributes: [

        {
          name: 'username',
          type: 'string',
          description: 'Username',
          pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
          minLength: 3,
          maxLength: 30,
          required: true
        },

        {
          name: 'first_name',
          type: 'string',
          description: 'First name',
          maxLength: 30
        },

        {
          name: 'last_name',
          type: 'string',
          description: 'Last name',
          maxLength: 30
        },

        {
          name: 'language',
          type: 'reference',
          description: 'User language',
          target: 'core::language',
          required: true
        },

        {
          name: 'email',
          type: 'email',
          description: 'Email address',
          required: true
        },

        {
          name: 'password',
          type: 'password',
          description: 'User password',
          required: true
        }
      ],

      indexing: {
        unique: [
          [ 'username' ],
          [ 'email' ],
        ]
      }

    }

  ]

}
