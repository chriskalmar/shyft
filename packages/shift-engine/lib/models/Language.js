
import Entity from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import Index, { INDEX_UNIQUE } from '../index/Index';


export const Language = new Entity({
  name: 'Language',
  description: 'A language',

  indexes: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'name' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'isoCode' ],
    }),
  ],

  attributes: {

    name: {
      type: DataTypeString,
      description: 'The name of the language',
      required: true
    },

    nativeName: {
      type: DataTypeString,
      description: 'The native name of the language',
      required: true
    },

    isoCode: {
      type: DataTypeString,
      description: 'ISO code of the language',
      required: true
    },
  }

})

