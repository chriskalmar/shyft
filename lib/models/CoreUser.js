
import {
  datatypes,
  Entity,
} from '../../';

import CoreLanguage from './CoreLanguage';

const {
  DataTypeString,
} = datatypes

export const CoreUser = new Entity({
  name: 'user',
  domain: 'core',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  // indexing: {
  //   unique: [
  //     [ 'username' ],
  //     [ 'email' ],


  attributes: {

    username: {
      type: DataTypeString,
      description: 'Username',
      pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
      minLength: 3,
      maxLength: 30,
      required: true
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
      maxLength: 30
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
      maxLength: 30
    },

    language: {
      type: CoreLanguage,
      description: 'User language',
      required: true
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true
    }
  }

})



