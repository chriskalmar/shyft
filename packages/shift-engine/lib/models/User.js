
import Entity from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import Index, { INDEX_UNIQUE } from '../index/Index';

import { Language } from './Language';


export const User = new Entity({
  name: 'User',
  domain: 'core',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  indexing: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'username' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'email' ],
    }),
  ],


  attributes: {

    username: {
      type: DataTypeString,
      description: 'Username',
      pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
      minLength: 3,
      maxLength: 30,
      required: true
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
      maxLength: 30
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
      maxLength: 30
    },

    language: {
      type: Language,
      description: 'User language',
      required: true
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true
    }
  }

})



