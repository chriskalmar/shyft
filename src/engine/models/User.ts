import { Entity } from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import { Index, INDEX_UNIQUE } from '../index/Index';

import { Language } from './Language';

export const User = new Entity({
  name: 'User',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  indexes: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'username' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'email' ],
    }),
  ],

  attributes: {
    username: {
      type: DataTypeString,
      description: 'Username',
      required: true,
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
    },

    language: {
      type: Language,
      description: 'User language',
      required: true,
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true,
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true,
    },
  },
});
