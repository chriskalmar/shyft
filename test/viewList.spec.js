import './setupAndTearDown';
import { find } from './db';

import { asAdmin, removeListDynamicData } from './testUtils';

import { Profile } from './models/Profile';
import { Board } from './models/Board';
import { BoardMemberView } from './models/BoardMemberView';

const orderByBoardIdAndInviterIdAsc = {
  orderBy: [
    {
      attribute: 'boardId',
      direction: 'ASC',
    },
    {
      attribute: 'inviterId',
      direction: 'ASC',
    },
  ],
};

const orderByBoardIdDesc = {
  orderBy: [
    {
      attribute: 'boardId',
      direction: 'DESC',
    },
  ],
};

describe('view list', () => {
  it('orderBy', async () => {
    const invites = await find(
      BoardMemberView,
      { ...orderByBoardIdAndInviterIdAsc },
      asAdmin(),
    );

    expect(invites).toMatchSnapshot();
  });

  it('orderBy + first', async () => {
    const invites = await find(
      BoardMemberView,
      { ...orderByBoardIdAndInviterIdAsc, first: 3 },
      asAdmin(),
    );

    expect(invites).toMatchSnapshot();
  });

  it('orderBy + first 0', async () => {
    const invites = await find(
      BoardMemberView,
      { ...orderByBoardIdAndInviterIdAsc, first: 0 },
      asAdmin(),
    );

    expect(invites).toMatchSnapshot();
  });

  it('orderBy + last', async () => {
    const invites = await find(
      BoardMemberView,
      { ...orderByBoardIdAndInviterIdAsc, last: 3 },
      asAdmin(),
    );

    expect(invites).toMatchSnapshot();
  });

  it('parentConnection', async () => {
    const parentConnection = {
      id: 79,
      attribute: 'inviterId',
    };

    const result = await find(
      BoardMemberView,
      { ...orderByBoardIdAndInviterIdAsc },
      asAdmin(),
      parentConnection,
    );

    expect(result).toMatchSnapshot();
  });

  it('parentConnection + first', async () => {
    const parentConnection = {
      id: 79,
      attribute: 'inviterId',
    };

    const result = await find(
      BoardMemberView,
      { ...orderByBoardIdAndInviterIdAsc, first: 1 },
      asAdmin(),
      parentConnection,
    );

    expect(result).toMatchSnapshot();
  });

  it('parentConnection with invalid attribute (reject)', async () => {
    const parentConnection = {
      id: 60,
      attribute: 'invalidAttributeName',
    };

    await find(
      BoardMemberView,
      { ...orderByBoardIdAndInviterIdAsc, first: 1 },
      asAdmin(),
      parentConnection,
    ).catch(e => {
      expect(e).toMatchSnapshot();
    });
  });
});
