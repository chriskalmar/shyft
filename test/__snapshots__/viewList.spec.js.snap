// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`view list descending orderBy + first 1`] = `
Object {
  "data": Array [
    BoardMemberView {
      "boardId": "50",
      "invitees": Array [
        "51",
        "4",
        "14",
        "90",
      ],
      "inviterId": "84",
    },
    BoardMemberView {
      "boardId": "49",
      "invitees": Array [
        "70",
        "8",
        "46",
        "100",
      ],
      "inviterId": "26",
    },
    BoardMemberView {
      "boardId": "48",
      "invitees": Array [
        "107",
        "81",
        "21",
        "103",
      ],
      "inviterId": "41",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPreviousPage": false,
  },
}
`;

exports[`view list descending orderBy + last 1`] = `
Object {
  "data": Array [
    BoardMemberView {
      "boardId": "1",
      "invitees": Array [
        "63",
      ],
      "inviterId": "63",
    },
    BoardMemberView {
      "boardId": "1",
      "invitees": Array [
        "42",
      ],
      "inviterId": "42",
    },
    BoardMemberView {
      "boardId": "1",
      "invitees": Array [
        "34",
      ],
      "inviterId": "34",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": false,
    "hasPreviousPage": true,
  },
}
`;

exports[`view list descending orderBy 1`] = `
Object {
  "data": Array [
    BoardMemberView {
      "boardId": "50",
      "invitees": Array [
        "51",
        "4",
        "14",
        "90",
      ],
      "inviterId": "84",
    },
    BoardMemberView {
      "boardId": "49",
      "invitees": Array [
        "70",
        "8",
        "46",
        "100",
      ],
      "inviterId": "26",
    },
    BoardMemberView {
      "boardId": "48",
      "invitees": Array [
        "107",
        "81",
        "21",
        "103",
      ],
      "inviterId": "41",
    },
    BoardMemberView {
      "boardId": "47",
      "invitees": Array [
        "104",
      ],
      "inviterId": "104",
    },
    BoardMemberView {
      "boardId": "47",
      "invitees": Array [
        "101",
      ],
      "inviterId": "101",
    },
    BoardMemberView {
      "boardId": "47",
      "invitees": Array [
        "70",
      ],
      "inviterId": "70",
    },
    BoardMemberView {
      "boardId": "47",
      "invitees": Array [
        "60",
      ],
      "inviterId": "60",
    },
    BoardMemberView {
      "boardId": "47",
      "invitees": Array [
        "58",
      ],
      "inviterId": "58",
    },
    BoardMemberView {
      "boardId": "47",
      "invitees": Array [
        "53",
      ],
      "inviterId": "53",
    },
    BoardMemberView {
      "boardId": "47",
      "invitees": Array [
        "49",
      ],
      "inviterId": "49",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPreviousPage": false,
  },
}
`;

exports[`view list orderBy + first 0 1`] = `
Object {
  "data": Array [],
  "pageInfo": Object {
    "hasNextPage": false,
    "hasPreviousPage": false,
  },
}
`;

exports[`view list orderBy + first 1`] = `
Object {
  "data": Array [
    BoardMemberView {
      "boardId": "1",
      "invitees": Array [
        "34",
      ],
      "inviterId": "34",
    },
    BoardMemberView {
      "boardId": "1",
      "invitees": Array [
        "42",
      ],
      "inviterId": "42",
    },
    BoardMemberView {
      "boardId": "1",
      "invitees": Array [
        "63",
      ],
      "inviterId": "63",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPreviousPage": false,
  },
}
`;

exports[`view list orderBy + last 1`] = `
Object {
  "data": Array [
    BoardMemberView {
      "boardId": "48",
      "invitees": Array [
        "107",
        "81",
        "21",
        "103",
      ],
      "inviterId": "41",
    },
    BoardMemberView {
      "boardId": "49",
      "invitees": Array [
        "70",
        "8",
        "46",
        "100",
      ],
      "inviterId": "26",
    },
    BoardMemberView {
      "boardId": "50",
      "invitees": Array [
        "51",
        "4",
        "14",
        "90",
      ],
      "inviterId": "84",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": false,
    "hasPreviousPage": true,
  },
}
`;

exports[`view list orderBy 1`] = `
Object {
  "data": Array [
    BoardMemberView {
      "boardId": "1",
      "invitees": Array [
        "34",
      ],
      "inviterId": "34",
    },
    BoardMemberView {
      "boardId": "1",
      "invitees": Array [
        "42",
      ],
      "inviterId": "42",
    },
    BoardMemberView {
      "boardId": "1",
      "invitees": Array [
        "63",
      ],
      "inviterId": "63",
    },
    BoardMemberView {
      "boardId": "1",
      "invitees": Array [
        "82",
      ],
      "inviterId": "82",
    },
    BoardMemberView {
      "boardId": "1",
      "invitees": Array [
        "91",
      ],
      "inviterId": "91",
    },
    BoardMemberView {
      "boardId": "2",
      "invitees": Array [
        "40",
        "92",
      ],
      "inviterId": "46",
    },
    BoardMemberView {
      "boardId": "3",
      "invitees": Array [
        "1",
        "38",
        "24",
        "61",
      ],
      "inviterId": "79",
    },
    BoardMemberView {
      "boardId": "4",
      "invitees": Array [
        "5",
      ],
      "inviterId": "5",
    },
    BoardMemberView {
      "boardId": "4",
      "invitees": Array [
        "9",
      ],
      "inviterId": "9",
    },
    BoardMemberView {
      "boardId": "4",
      "invitees": Array [
        "17",
      ],
      "inviterId": "17",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPreviousPage": false,
  },
}
`;

exports[`view list parentConnection + first 1`] = `
Object {
  "data": Array [
    BoardMemberView {
      "boardId": "3",
      "invitees": Array [
        "24",
        "38",
        "1",
        "61",
      ],
      "inviterId": "79",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPreviousPage": false,
  },
}
`;

exports[`view list parentConnection 1`] = `
Object {
  "data": Array [
    BoardMemberView {
      "boardId": "3",
      "invitees": Array [
        "24",
        "38",
        "1",
        "61",
      ],
      "inviterId": "79",
    },
    BoardMemberView {
      "boardId": "13",
      "invitees": Array [
        "79",
      ],
      "inviterId": "79",
    },
    BoardMemberView {
      "boardId": "25",
      "invitees": Array [
        "79",
      ],
      "inviterId": "79",
    },
    BoardMemberView {
      "boardId": "29",
      "invitees": Array [
        "79",
      ],
      "inviterId": "79",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": false,
    "hasPreviousPage": false,
  },
}
`;

exports[`view list parentConnection with invalid attribute (reject) 1`] = `[Error: column "invalidAttributeName" does not exist]`;
